library/core/src/slice/mod.rs

```diff
@@ -1843,25 +1843,24 @@ impl<T> [T] {
     where
         F: FnMut(&'a T) -> Ordering,
     {
-        let s = self;
-        let mut size = s.len();
-        if size == 0 {
-            return Err(0);
-        }
-        let mut base = 0usize;
-        while size > 1 {
-            let half = size / 2;
-            let mid = base + half;
-            // SAFETY: the call is made safe by the following inconstants:
-            // - `mid >= 0`: by definition
-            // - `mid < size`: `mid = size / 2 + size / 4 + size / 8 ...`
-            let cmp = f(unsafe { s.get_unchecked(mid) });
-            base = if cmp == Greater { base } else { mid };
-            size -= half;
+        let mut left = 0;
+        let mut right = self.len();
+        while left < right {
+            // never overflow because `slice::len()` max is `isize::MAX`.
```

Comment:
Is this guaranteed for slices in general? Or only for `Vec`.

Reply:
It's only true for slices of non-ZSTs, so good point.  This needs further elaboration for the ZST case.

(Personally I'd suggest rewriting it to the `let mid = base + half;` approach that was previously there.  Last I played with a binary search, LLVM didn't like the `let mid = (left + right) / 2;` approach -- it looked like it couldn't canonicalize the loop, perhaps because it gives it two loop variables [left + right] rather than one [size].)

Reply:
Hi @scottmcm. Thanks for reviewing. Could you please explain why we need further elaboration for the ZST case? I have no idea about this.

Reply:
The code says "// never overflow because `slice::len()` max is `isize::MAX`."

That's correct for non-ZSTs, but strictly the restriction, as seen in https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety is that the length *in bytes* is at most that -- so if `sizeof(T) == 0` then the size in bytes is also `0`, and `usize::MAX` is allowed.  Here's a quick demo: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e406602a8f80fda028463f0c8cb1dd22

Reply:
Would it be sensible to just return Ok(0) or Err(0) as all ZSTs shall be equal?